File: backend/app/services/batch_audit_service.py
"""
Batch Audit Service - Processamento em Lote com Rate Limiting
"""
import asyncio
from typing import List, Dict, Any
from datetime import datetime
from pydantic import BaseModel
from app.services.cnpj_service import CNPJService
from app.services.cnpj.base import CNPJRateLimitError, CNPJAPIError

class BatchAuditRequest(BaseModel):
    """Request para auditoria em lote"""
    items: List[Dict[str, Any]]  # Lista de {cnpj, transaction_id, service_type}

class BatchAuditStatus(BaseModel):
    """Status do processamento em lote"""
    total: int
    processed: int
    pending: int
    results: List[Dict[str, Any]]
    errors: List[Dict[str, Any]]

class BatchAuditService:
    """
    Serviço de auditoria em lote com rate limiting inteligente.
    Respeita limite de 3 req/min da API grátis.
    """
    
    def __init__(self):
        self.cnpj_service = CNPJService()
        self.results = []
        self.errors = []
    
    async def process_batch(
        self,
        items: List[Dict[str, Any]],
        progress_callback=None
    ) -> BatchAuditStatus:
        """
        Processa lista de CNPJs em lote.
        
        Args:
            items: Lista de dicts com {cnpj, transaction_id, service_type}
            progress_callback: Função para reportar progresso
        """
        total = len(items)
        processed = 0
        
        print(f"[Batch Audit] Iniciando processamento de {total} itens...")
        
        for i, item in enumerate(items):
            try:
                # Processar item
                result = await self._process_single_item(item)
                self.results.append(result)
                processed += 1
                
                # Reportar progresso
                if progress_callback:
                    progress_callback(processed, total)
                
                print(f"[Batch Audit] {processed}/{total} - {item.get('cnpj')} ✅")
                
                # Rate limiting: Aguardar entre requests (versão grátis)
                if not self.cnpj_service.provider.is_paid and i < total - 1:
                    wait_time = 20  # 20 segundos = 3 req/min
                    print(f"[Batch Audit] Aguardando {wait_time}s (rate limit)...")
                    await asyncio.sleep(wait_time)
                
            except CNPJRateLimitError as e:
                # Rate limit atingido - aguardar mais tempo
                print(f"[Batch Audit] Rate limit atingido! Aguardando 60s...")
                await asyncio.sleep(60)
                
                # Tentar novamente
                try:
                    result = await self._process_single_item(item)
                    self.results.append(result)
                    processed += 1
                except Exception as retry_error:
                    self.errors.append({
                        "item": item,
                        "error": str(retry_error)
                    })
                    
            except Exception as e:
                print(f"[Batch Audit] Erro ao processar {item.get('cnpj')}: {str(e)}")
                self.errors.append({
                    "item": item,
                    "error": str(e)
                })
        
        pending = total - processed
        
        return BatchAuditStatus(
            total=total,
            processed=processed,
            pending=pending,
            results=self.results,
            errors=self.errors
        )
    
    async def _process_single_item(self, item: Dict[str, Any]) -> Dict[str, Any]:
        """Processa um único item"""
        cnpj = item.get("cnpj")
        service_type = item.get("service_type", "")
        
        # Validar CNPJ
        supplier_data = await self.cnpj_service.validate_cnpj(cnpj)
        
        # Determinar risco
        risk_level = self.cnpj_service.get_risk_level(supplier_data)
        
        # Validar CNAE vs Serviço
        from app.services.robust_validator import RobustValidator
        validator = RobustValidator()
        
        cnae_valid, cnae_reason = validator.validate_cnae_service(
            cnae_principal=supplier_data.cnae_principal.codigo,
            cnaes_secundarios=[],  # TODO: Pegar CNAEs secundários da API
            service_type=service_type
        )
        
        # Determinar status final
        if risk_level == "CRITICAL_RISK":
            status = "REJECTED"
            reason = f"Empresa {supplier_data.status_receita}"
        elif cnae_valid == False:
            status = "CNAE_MISMATCH"
            reason = cnae_reason
        elif cnae_valid == None:
            status = "MANUAL_REVIEW"
            reason = cnae_reason
        else:
            status = "APPROVED"
            reason = "Fornecedor validado"
        
        return {
            "cnpj": cnpj,
            "razao_social": supplier_data.razao_social,
            "status_receita": supplier_data.status_receita,
            "cnae": supplier_data.cnae_principal.codigo,
            "risk_level": risk_level,
            "cnae_valid": cnae_valid,
            "status": status,
            "reason": reason,
            "cached": supplier_data.cached
        }

File: supabase/functions/_shared/batch-audit-service.ts
/**
 * Batch Audit Service - TypeScript
 * Portado de: backend/app/services/batch_audit_service.py
 * 
 * Processamento em lote com rate limiting inteligente.
 */

import { CNPJService, RiskLevel } from "./cnpj-service.ts";
import { RobustValidator } from "./robust-validator.ts";

declare const Deno: any;

// ============== INTERFACES ==============

export interface BatchAuditItem {
    cnpj: string;
    transaction_id?: string;
    service_type?: string;
}

export interface BatchAuditResult {
    cnpj: string;
    razao_social: string;
    status_receita: string;
    cnae: string;
    risk_level: RiskLevel;
    cnae_valid: boolean | null;
    status: "APPROVED" | "REJECTED" | "CNAE_MISMATCH" | "MANUAL_REVIEW";
    reason: string;
    cached: boolean;
}

export interface BatchAuditStatus {
    total: number;
    processed: number;
    pending: number;
    results: BatchAuditResult[];
    errors: Array<{
        item: BatchAuditItem;
        error: string;
    }>;
}

// ============== BATCH AUDIT SERVICE ==============

export class BatchAuditService {
    private cnpjService: CNPJService;
    private results: BatchAuditResult[] = [];
    private errors: Array<{ item: BatchAuditItem; error: string }> = [];

    constructor() {
        this.cnpjService = new CNPJService();
    }

    /**
     * Processa lista de CNPJs em lote com rate limiting.
     */
    async processBatch(
        items: BatchAuditItem[],
        progressCallback?: (processed: number, total: number) => void
    ): Promise<BatchAuditStatus> {
        const total = items.length;
        let processed = 0;

        console.log(`[Batch Audit] Iniciando processamento de ${total} itens...`);

        for (let i = 0; i < items.length; i++) {
            const item = items[i];

            try {
                // Processar item
                const result = await this.processSingleItem(item);
                this.results.push(result);
                processed++;

                // Reportar progresso
                if (progressCallback) {
                    progressCallback(processed, total);
                }

                console.log(`[Batch Audit] ${processed}/${total} - ${item.cnpj} ✅`);

                // Rate limiting para versão grátis CNPJ.ws (3 req/min = 20s entre requests)
                if (i < total - 1) {
                    const waitTime = 20000; // 20 segundos
                    console.log(`[Batch Audit] Aguardando ${waitTime / 1000}s (rate limit)...`);
                    await new Promise(resolve => setTimeout(resolve, waitTime));
                }

            } catch (error: any) {
                if (error.name === 'CNPJRateLimitError') {
                    // Rate limit atingido - aguardar mais tempo
                    console.log(`[Batch Audit] Rate limit atingido! Aguardando 60s...`);
                    await new Promise(resolve => setTimeout(resolve, 60000));

                    // Tentar novamente
                    try {
                        const result = await this.processSingleItem(item);
                        this.results.push(result);
                        processed++;
                    } catch (retryError: any) {
                        this.errors.push({
                            item,
                            error: retryError.message
                        });
                    }
                } else {
                    console.log(`[Batch Audit] Erro ao processar ${item.cnpj}: ${error.message}`);
                    this.errors.push({
                        item,
                        error: error.message
                    });
                }
            }
        }

        const pending = total - processed;

        return {
            total,
            processed,
            pending,
            results: this.results,
            errors: this.errors
        };
    }

    /**
     * Processa um único item
     */
    private async processSingleItem(item: BatchAuditItem): Promise<BatchAuditResult> {
        const cnpj = item.cnpj;
        const serviceType = item.service_type || "";

        // Validar CNPJ
        const supplierData = await this.cnpjService.validateCNPJ(cnpj);

        // Determinar risco
        const riskLevel = this.cnpjService.getRiskLevel(supplierData);

        // Validar CNAE vs Serviço
        const validator = new RobustValidator();
        const cnaeValidation = validator.validateCnaeService(
            supplierData.cnae_principal.codigo,
            [],
            serviceType
        );

        // Determinar status final
        let status: BatchAuditResult['status'];
        let reason: string;

        if (riskLevel === "CRITICAL_RISK") {
            status = "REJECTED";
            reason = `Empresa ${supplierData.status_receita}`;
        } else if (cnaeValidation.valid === false) {
            status = "CNAE_MISMATCH";
            reason = cnaeValidation.reason;
        } else if (cnaeValidation.valid === null) {
            status = "MANUAL_REVIEW";
            reason = cnaeValidation.reason;
        } else {
            status = "APPROVED";
            reason = "Fornecedor validado";
        }

        return {
            cnpj,
            razao_social: supplierData.razao_social,
            status_receita: supplierData.status_receita,
            cnae: supplierData.cnae_principal.codigo,
            risk_level: riskLevel,
            cnae_valid: cnaeValidation.valid,
            status,
            reason,
            cached: supplierData.cached
        };
    }

    /**
     * Limpa resultados e erros para novo processamento
     */
    reset(): void {
        this.results = [];
        this.errors = [];
    }
}
